sdk/core/core/inc/az_json.h

```diff
@@ -182,81 +196,233 @@ AZ_NODISCARD az_result az_json_token_get_string(az_json_token const* token, az_s
+ *         - #AZ_OK if the number was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_number(az_json_builder* json_builder, double value);
+
+/**
+ * @brief Appends an int32_t number value.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the number to.
+ * @param[in] value The value to be written as a JSON number.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the number was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
  */
 AZ_NODISCARD az_result
-az_json_builder_append_array_item(az_json_builder* json_builder, az_json_token token);
+az_json_builder_append_int32_number(az_json_builder* json_builder, int32_t value);
+
+/**
+ * @brief Appends the JSON literal `null`.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the `null` literal to.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if `null` was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_null(az_json_builder* json_builder);
+
+/**
+ * @brief Appends the beginning of a JSON object (i.e. `{`).
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the start of object to.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if object start was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_object_start(az_json_builder* json_builder);
+
+/**
+ * @brief Appends the beginning of a JSON array (i.e. `[`).
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the start of array to.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if array start was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_array_start(az_json_builder* json_builder);
+
+/**
+ * @brief Appends the end of the current JSON object (i.e. `}`) or JSON array (i.e. `]`), depending
+ * on which JSON container is currently open.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance containing the buffer to append
+ * the closing character to.
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if container end was appended successfully
+ *         - #AZ_ERROR_INSUFFICIENT_SPAN_SIZE if the buffer is too small
+ */
+AZ_NODISCARD az_result az_json_builder_append_container_end(az_json_builder* json_builder);
```

Comment:
Maybe rename to *_append_array_or_object_end? 
This way the function name has some relationship to the *_start function.
Alternatively, what if you had *_append_container_start which took an AZ_TOKEN_OBJECT_START or AZ_TOKEN_ARRAY_START parameter and then you keep *_container_end.

Reply:
Both are reasonable. I was a bit weary of introducing a new term that isn't really part of the domain (container) so I do like `_append_array_or_object_end`. The only concern is with that the method name becomes super long/verbose.

Is just `az_json_builder_append_end` too ambiguous? It mirrors the `*_start` methods, at least around the suffix.

Reply:
I would just call it `az_json_builder_close_current_object`.  a json object can be either [ ] or { }. Both are json valid objects.
I would also have an useful function `az_json_builder_closee_all_remaining` to terminate the json with all required closings

Reply:
*_close_object (i would drop "current") isn't bad but then we're switching from start/end to open/close. I'm OK with the switch but consistency is important. I thin open/close is better than start/end because I always say "close brace" or "open square bracket" when I speak of these things.

Reply:
> I would just call it `az_json_builder_close_current_object`.  a json object can be either [ ] or { }. Both are json valid objects.

Let's make sure to use the correct terminology. A JSON object is **NOT** either [] or {}. They are distinct things. One is a JSON array and the other is a JSON object.

I don't think we should say `_close_object` to refer to both.

> I would also have an useful function `az_json_builder_closee_all_remaining` to terminate the json with all required closings

Maybe that could be useful, but I don't see that being all that common. The caller can very easily write their own helper function on top if they need that behavior.

> I thin open/close is better than start/end because I always say "close brace" or "open square bracket" when I speak of these things.

Another option is begin/end (which is probably where start/end came from) since that is what the RFC uses.